### PR TITLE
Replace (Par)TrieMap with ConcurrentHashMap in IncOptimizer

### PR DESCRIPTION
**Pull request by gzm0**
- Reduces residual retained size of the IncOptimizer on the test suite
  from 166MB to 144MB.
- ~10% batch speedup on the optimizer.

### CODE REVIEW

**Comment by gzm0** [comment]
plot

[logger-timings.csv](https://github.com/scala-js/scala-js/files/14886851/logger-timings.csv)

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala [hunk 1]
```diff
@@ -65,15 +76,22 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
...
 
   private var methodsToProcess = collOps.emptyAddable[Processable]
 
   @inline
   private def getInterface(className: ClassName): InterfaceType =
-    collOps.forceGet(interfaces, className)
+    interfaces.get(className)
+
+  @inline
+  private def classOrElse[T >: Class](className: ClassName, default: =>T): T = {
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:90]: Coding style nit:
```suggestion
  private def classOrElse[T >: Class](className: ClassName, default: => T): T = {
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Very nice :)

I only have one real comment about the tests.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala [hunk 2]
```diff
@@ -227,6 +227,63 @@ class ConcurrentHashMapTest extends MapTest {
     val str = keySet.toString
     assertTrue(s"toString should print keys, but actual: $str", str == "[a, b]" || str == "[b, a]")
   }
+
+  @Test def forEachPar(): Unit = {
+    val pairs = List("ONE" -> 1, "TWO" -> 2, "THREE" -> 3)
+    val map = factory.empty[String, Int]
+    pairs.foreach(x => map.put(x._1, x._2))
+
+    val seen = mutable.Set.empty[(String, Int)]
```
[the remark this answers is not in the document]
Comment (gzm0) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala:236]: Good catch.

**Comment by gzm0** [comment]
I was having second thoughts about implementing the supposedly parallel `forEach` serially unconditionally. I have added a commit (to be squashed) which disallows values of `parallelismThreshold` other than `Long.MaxValue`.

WDYT?

**Comment by sjrd** [comment]
Hum I don't think that's OK. It directly goes against our policy "if it links, it's correct". Does the API *guarantee* any degree of parallelism? It didn't seem like it does when I looked at it. If it doesn't, always executing serially is correct.

**Comment by gzm0** [comment]
JavaDoc says:

> Using a value of 1 results in maximal parallelism by partitioning into enough subtasks to fully utilize the [ForkJoinPool.commonPool()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#commonPool--) that is used for all parallel computations.

So yeah, we could argue that if we ever were to implement `ForkJoinPool.commonPool`, we'd implement `ForkJoinPool.getCommonPoolParallelism` to return 1. So that in turn would mean it's OK to allow any parallelism level.

**Comment by gzm0** [comment]
I have removed the last commit again.

**Review by sjrd — APPROVED** [review]

**Pull request merged**